src/Workspaces/Core/Portable/Workspace/Solution/StateChecksums.cs

```diff
@@ -102,39 +121,48 @@ public static SolutionStateChecksums Deserialize(ObjectReader reader)
 
         ChecksumCollection.Find(state.AnalyzerReferences, AnalyzerReferences, searchingChecksumsLeft, result, cancellationToken);
 
-        // Before doing a depth-first-search *into* each project, first run across all the project at their top level.
-        // This ensures that when we are trying to sync the projects referenced by a SolutionStateChecksums' instance
-        // that we don't unnecessarily walk all documents looking just for those.
+        if (searchingChecksumsLeft.Count == 0)
+            return;
 
-        foreach (var (projectId, projectState) in state.ProjectStates)
+        if (hintProject != null)
         {
-            if (searchingChecksumsLeft.Count == 0)
-                break;
-
-            if (hintProject != null && hintProject != projectId)
-                continue;
-
-            if (projectState.TryGetStateChecksums(out var projectStateChecksums) &&
-                searchingChecksumsLeft.Remove(projectStateChecksums.Checksum))
+            var projectState = state.GetProjectState(hintProject);
+            if (projectState != null &&
+                projectState.TryGetStateChecksums(out var projectStateChecksums))
             {
```

Comment:
a searchingChecksumsLeft.Remove call isn't required here?

Reply:
no.  this isn't removing.  this is just fast-pathing finding the project-state-checksums to dive into it to search the actual checksums.